save-demo/src/main/kotlin/com/saveourtool/save/demo/service/KubernetesService.kt

```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * @param demo demo entity
+     * @param version version of [demo] that should be run in kubernetes pod
+     * @return [Mono] of [StringResponse] filled with readable message
+     */
+    @Suppress("TOO_MANY_LINES_IN_LAMBDA")
+    fun start(demo: Demo, version: String = "manual"): Mono<StringResponse> = Mono.just(demo)
+        .logValue(logger::info) {
+            "Creating job ${jobNameForDemo(it)}..."
+        }
+        .flatMap { requestedDemo ->
+            try {
+                requestedDemo.also { kc.startJob(requestedDemo, kubernetesSettings) }.toMono()
+            } catch (kre: KubernetesRunnerException) {
+                Mono.error(
+                    ResponseStatusException(
+                        HttpStatus.INTERNAL_SERVER_ERROR,
+                        "Could not create job for ${jobNameForDemo(requestedDemo)}",
+                        kre,
+                    )
+                )
+            }
+        }
+        .flatMap { requestedDemo ->
+            deferredToMono {
+                configureDemoAgent(requestedDemo, version)
+            }
+        }
+
+    /**
+     * @param demo demo entity
+     * @return list of [StatusDetails]
+     */
+    fun stop(demo: Demo): List<StatusDetails> {
+        logger.info("Stopping job...")
+        return kc.getJob(demo, configProperties.kubernetes.namespace).delete()
+    }
+
+    /**
+     * @param demo demo entity
+     * @return deferred filled with [DemoStatus]
+     */
+    fun getStatus(demo: Demo) = scope.async {
+        val url = "${getPodUrl(demo)}/alive"
+        val status = try {
+            httpClient.get(url).status
+        } catch (connectException: ConnectException) {
+            null
+        }
+        when {
+            status == null -> DemoStatus.ERROR
+            status.isSuccess() -> DemoStatus.RUNNING
+            else -> DemoStatus.STOPPED
+        }
+    }
+
+    private fun configureDemoAgent(demo: Demo, version: String) = scope.async {
+        val url = "${getPodUrl(demo)}/configure"
+        logger.info("Configuring save-demo-agent by url: $url")
+        val requestStatus = httpClient.post(url) {
+            contentType(ContentType.Application.Json)
+            accept(ContentType.Application.Json)
+            setBody(
+                DemoAgentConfig(
+                    configProperties.agentConfig.demoUrl,
+                    demo.toDemoConfiguration(version),
+                    demo.toRunConfiguration(),
+                )
+            )
+        }.status
+
+        if (requestStatus.isSuccess()) {
+            logAndRespond(HttpStatusCode.OK, logger::info) {
+                "Job successfully started."
+            }
+        } else {
+            logAndRespond(HttpStatusCode.InternalServerError, logger::error) {
+                "Could not configure demo."
+            }
+        }
+    }
+
+    private suspend fun getPodUrl(demo: Demo) = retry(RETRY_TIMES) {
+        kc.getJobPodsIps(demo).firstOrNull()
```

Comment:
since you use only -- save-agent-demo runs on :80?

also, `.firstOrNull()` -- do we are going to support a several pods for a single demo?

Reply:
Once, I guess, yes